File: NetDeck/FilterViewController.m
//
//  FilterViewController.m
//  Net Deck
//
//  Created by Gereon Steffens on 24.08.15.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "FilterViewController.h"

@interface FilterViewController ()

@property NSArray* factionNames;
@property NSArray* typeNames;
@property BOOL dataDestinyAllowed;

@property NSArray* cards;
@property BOOL showPreviewTable;

@end

@implementation FilterViewController 

enum { TAG_FACTION, TAG_MINI_FACTION, TAG_TYPE };

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"Filter";
    
    self.dataDestinyAllowed = [[NSUserDefaults standardUserDefaults] boolForKey:SettingsKeys.USE_DATA_DESTINY];
    self.factionLabel.text = l10n(@"Faction");
    self.typeLabel.text = l10n(@"Type");
    
    self.typeVerticalDistance.constant = 18;
    self.miniFactionControl.hidden = YES;
    self.miniFactionControl.tag = TAG_MINI_FACTION;
    self.miniFactionControl.delegate = self;
    [self.miniFactionControl selectAllSegments:self.dataDestinyAllowed];
    
    NSInteger factionLimit;
    if (self.role == NRRoleRunner)
    {
        self.factionNames = @[ [Faction name:NRFactionAnarch], [Faction name:NRFactionCriminal], [Faction name:NRFactionShaper], [Faction name:NRFactionNeutral]];
        self.typeNames = @[ [CardType name:NRCardTypeEvent], [CardType name:NRCardTypeHardware], [CardType name:NRCardTypeResource], [CardType name:NRCardTypeProgram] ];
        factionLimit = self.factionNames.count;
        
        if (self.dataDestinyAllowed)
        {
            self.factionNames = @[ [Faction name:NRFactionAnarch], [Faction name:NRFactionCriminal], [Faction name:NRFactionShaper], [Faction name:NRFactionNeutral],
                                   [Faction name:NRFactionAdam], [Faction name:NRFactionApex], [Faction name:NRFactionSunnyLebeau] ];
            
            self.miniFactionControl.hidden = NO;
            self.typeVerticalDistance.constant = 50;
        }
    }
    else
    {
        self.factionNames = @[ [Faction name:NRFactionHaasBioroid], [Faction name:NRFactionNBN], [Faction name:NRFactionJinteki], [Faction name:NRFactionWeyland], [Faction name:NRFactionNeutral]];
        factionLimit = self.factionNames.count;
        
        self.typeNames = @[ [CardType name:NRCardTypeAgenda], [CardType name:NRCardTypeAsset], [CardType name:NRCardTypeUpgrade], [CardType name:NRCardTypeOperation], [CardType name:NRCardTypeIce]];
    }
    
    self.factionControl.delegate = self;
    self.factionControl.tag = TAG_FACTION;
    [self.factionControl removeAllSegments];
    
    for (NSInteger i = 0; i<factionLimit; ++i)
    {
        [self.factionControl insertSegmentWithTitle:self.factionNames[i] atIndex:i animated:NO];
    }
    [self.factionControl selectAllSegments:YES];
    
    self.typeControl.delegate = self;
    self.typeControl.tag = TAG_TYPE;
    [self.typeControl removeAllSegments];
    
    for (NSInteger i = 0; i<self.typeNames.count; ++i)
    {
        [self.typeControl insertSegmentWithTitle:self.typeNames[i] atIndex:i animated:NO];
    }
    [self.typeControl selectAllSegments:YES];
    
    self.costSlider.maximumValue = 1+(self.role == NRRoleRunner ? [CardManager maxRunnerCost] : [CardManager maxCorpCost]);
    self.muApSlider.maximumValue = self.role == NRRoleRunner ? 1+[CardManager maxMU] : 1+[CardManager maxAgendaPoints];
    self.strengthSlider.maximumValue = 1+[CardManager maxStrength];
    self.influenceSlider.maximumValue = 1+[CardManager maxInfluence];
    
    [self.costSlider setThumbImage:[UIImage imageNamed:@"credit_slider"] forState:UIControlStateNormal];
    [self.muApSlider setThumbImage:[UIImage imageNamed:self.role == NRRoleRunner ? @"mem_slider" : @"point_slider" ] forState:UIControlStateNormal];
    [self.strengthSlider setThumbImage:[UIImage imageNamed:@"strength_slider"] forState:UIControlStateNormal];
    [self.influenceSlider setThumbImage:[UIImage imageNamed:@"influence_slider"] forState:UIControlStateNormal];
    
    [self clearFilters:nil];
    
    self.previewTable.rowHeight = 30;
    self.previewTable.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.showPreviewTable = YES;
    
    if (self.parentViewController.view.frame.size.height == 480)
    {
        // iphone 4s
        self.previewTable.scrollEnabled = NO;
        self.showPreviewTable = NO;
    }
    
    self.previewHeader.font = [UIFont md_systemFontOfSize:15];
}

-(void) viewDidAppear:(BOOL)animated
{
    NSAssert(self.navigationController.viewControllers.count == 4, @"nav oops");
    
    UINavigationItem* topItem = self.navigationController.navigationBar.topItem;
    UIBarButtonItem* clearButton = [[UIBarButtonItem alloc] initWithTitle:l10n(@"Clear") style:UIBarButtonItemStylePlain target:self action:@selector(clearFilters:)];
    topItem.rightBarButtonItem = clearButton;
    
    [super viewDidAppear:animated];
}

-(void) clearFilters:(id)sender
{
    [self.cardList clearFilters];
    
    [self.factionControl selectAllSegments:YES];
    [self.miniFactionControl selectAllSegments:self.dataDestinyAllowed];
    [self.typeControl selectAllSegments:YES];
    
    self.influenceSlider.value = 0;
    self.strengthSlider.value = 0;
    self.muApSlider.value = 0;
    self.costSlider.value = 0;
    
    [self influenceChanged:nil];
    [self strengthChanged:nil];
    [self muApChanged:nil];
    [self costChanged:nil];
}

-(void) strengthChanged:(UISlider*)sender
{
    int value = round(sender.value);
    // NSLog(@"str: %f %d", sender.value, value);
    sender.value = value--;
    self.strengthLabel.text = [NSString stringWithFormat:l10n(@"Strength: %@"), value == -1 ? l10n(@"All") : [@(value) stringValue]];
    [self.cardList filterByStrength:value];
    [self updatePreview];
}

-(void) muApChanged:(UISlider*)sender
{
    int value = round(sender.value);
    // NSLog(@"mu: %f %d", sender.value, value);
    sender.value = value--;
    self.muApLabel.text = [NSString stringWithFormat:self.role == NRRoleRunner ? l10n(@"MU: %@") : l10n(@"AP: %@"), value == -1 ? l10n(@"All") : [@(value) stringValue]];
    
    if (self.role == NRRoleRunner)
    {
        [self.cardList filterByMU:value];
    }
    else
    {
        [self.cardList filterByAgendaPoints:value];
    }
    [self updatePreview];
}

-(void) costChanged:(UISlider*)sender
{
    int value = round(sender.value);
    // NSLog(@"cost: %f %d", sender.value, value);
    sender.value = value--;
    self.costLabel.text = [NSString stringWithFormat:l10n(@"Cost: %@"), value == -1 ? l10n(@"All") : [@(value) stringValue]];
    [self.cardList filterByCost:value];
    [self updatePreview];
}

-(void) influenceChanged:(UISlider*)sender
{
    int value = round(sender.value);
    // NSLog(@"inf: %f %d", sender.value, value);
    sender.value = value--;
    self.influenceLabel.text = [NSString stringWithFormat:l10n(@"Influence: %@"), value == -1 ? l10n(@"All") : [@(value) stringValue]];
    
    if (self.identity)
    {
        [self.cardList filterByInfluence:value forFaction:self.identity.faction];
    }
    else
    {
        [self.cardList filterByInfluence:value];
    }
    [self updatePreview];
}

#pragma mark - multi select delegate

-(void)multiSelect:(MultiSelectSegmentedControl *)control didChangeValue:(BOOL)value atIndex:(NSUInteger)index
{
    NSArray* values = control.tag == TAG_TYPE ? self.typeNames : self.factionNames;
    
    NSMutableSet* set = [NSMutableSet set];

    if (control.tag == TAG_TYPE)
    {
        [control.selectedSegmentIndexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
            [set addObject:values[idx]];
        }];
        [self.cardList filterByTypes:set];
    }
    else
    {
        [self.factionControl.selectedSegmentIndexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
            [set addObject:values[idx]];
        }];
        if (self.role == NRRoleRunner)
        {
            [self.miniFactionControl.selectedSegmentIndexes enumerateIndexesUsingBlock:^(NSUInteger idx, BOOL *stop) {
                [set addObject:values[idx+4]];
            }];
        }
        
        [self.cardList filterByFactions:set];
    }
    [self updatePreview];
}

#pragma mark - table view

-(void) updatePreview
{
    self.cards = self.cardList.allCards;
    
    NSInteger count = self.cards.count;
    NSString* fmt = count == 1 ? l10n(@"%lu matching card") : l10n(@"%lu matching cards");
    NSString* text = @"  ";
    text = [text stringByAppendingString:[NSString stringWithFormat:fmt, count]];
    self.previewHeader.text = text;
    
    [self.previewTable reloadData];
}

-(NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.showPreviewTable ? self.cards.count : 0;
}

-(UITableViewCell*) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString* cellIdentifier = @"previewCell";
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (!cell)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
        cell.selectionStyle = UITableViewCellSelectionStyleNone;
        cell.textLabel.font = [UIFont systemFontOfSize:13];
    }
    
    Card* card = self.cards[indexPath.row];
    cell.textLabel.text = card.name;
    
    return cell;
}

@end

File: NetDeck/FilterViewController.swift
//
//  FilterViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 11.12.16.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

import UIKit
import MultiSelectSegmentedControl
import SwiftyUserDefaults

private enum Tags: Int {
    case faction
    case miniFaction
    case type
}

class FilterViewController: UIViewController, MultiSelectSegmentedControlDelegate, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet weak var factionControl: MultiSelectSegmentedControl!
    @IBOutlet weak var miniFactionControl: MultiSelectSegmentedControl!
    
    @IBOutlet weak var typeVerticalDistance: NSLayoutConstraint!
    @IBOutlet weak var typeControl: MultiSelectSegmentedControl!
    
    @IBOutlet weak var influenceLabel: UILabel!
    @IBOutlet weak var influenceSlider: UISlider!
    
    @IBOutlet weak var strengthLabel: UILabel!
    @IBOutlet weak var strengthSlider: UISlider!
    
    @IBOutlet weak var muApLabel: UILabel!
    @IBOutlet weak var muApSlider: UISlider!
    
    @IBOutlet weak var costLabel: UILabel!
    @IBOutlet weak var costSlider: UISlider!
    
    @IBOutlet weak var previewTable: UITableView!
    @IBOutlet weak var previewHeader: UILabel!
    
    var role = Role.none
    var identity: Card?
    var cardList: CardList!
    
    private var factionNames = [String]()
    private var typeNames = [String]()
    
    private var selectedTypes = Set<String>()
    private var subtypes: FilterValue?
    
    private var cards = [Card]()
    private var showPreviewTable = true

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "Filter".localized()
                
        self.typeVerticalDistance.constant = 16
        self.miniFactionControl.isHidden = true
        self.miniFactionControl.tag = Tags.miniFaction.rawValue
        self.miniFactionControl.delegate = self
        self.miniFactionControl.selectAllSegments(false)
        
        let useDaD = Defaults[.useDataDestiny]
        let packUsage = Defaults[.deckbuilderPacks]
        let miniFactions = packUsage == .all || useDaD
        
        let factionLimit: Int
        if self.role == .runner {
            self.factionNames = (Faction.runnerFactionsCore + [ Faction.neutral ]).map { Faction.name(for: $0) }
            self.typeNames = Array(CardType.runnerTypeNames.dropFirst()) // remove "Any"
            factionLimit = factionNames.count
            if miniFactions {
                self.factionNames = (Faction.runnerFactionsCore + [ Faction.neutral ] + Faction.runnerMiniFactions).map { Faction.name(for: $0) }
                self.miniFactionControl.isHidden = false
                self.typeVerticalDistance.constant = 48
            }
        } else {
            self.factionNames = (Faction.corpFactions + [ Faction.neutral ]).map { Faction.name(for: $0) }
            factionLimit = factionNames.count
            self.typeNames = Array(CardType.corpTypeNames.dropFirst()) // remove "Any"
        }
        
        // faction control
        self.factionControl.delegate = self
        self.factionControl.tag = Tags.faction.rawValue
        
        self.factionControl .removeAllSegments()
        for i in 0 ..< factionLimit {
            self.factionControl.insertSegment(withTitle: self.factionNames[i], at: i, animated: false)
        }
        self.factionControl.selectAllSegments(false)
        
        // type control
        self.typeControl.delegate = self
        self.typeControl.tag = Tags.type.rawValue
        
        self.typeControl.removeAllSegments()
        for i in 0 ..< self.typeNames.count {
            self.typeControl.insertSegment(withTitle: self.typeNames[i], at: i, animated: false)
        }
        self.typeControl.selectAllSegments(false)
        
        // sliders
        self.costSlider.maximumValue = Float(1 + (self.role == .runner ? CardManager.maxRunnerCost : CardManager.maxCorpCost))
        self.muApSlider.maximumValue = Float(1 + (self.role == .runner ? CardManager.maxMU : CardManager.maxAgendaPoints))
        self.strengthSlider.maximumValue = Float(1 + CardManager.maxStrength)
        self.influenceSlider.maximumValue = Float(1 + CardManager.maxInfluence)
        
        self.costSlider.setThumbImage(UIImage(named: "credit_slider"), for: .normal)
        let iconName = self.role == .runner ? "mem_slider" : "point_slider"
        self.muApSlider.setThumbImage(UIImage(named: iconName), for: .normal)
        self.strengthSlider.setThumbImage(UIImage(named: "strength_slider"), for: .normal)
        self.influenceSlider.setThumbImage(UIImage(named: "influence_slider"), for: .normal)
        
        self.clearFilters()
        
        self.previewHeader.font = UIFont.monospacedDigitSystemFont(ofSize: 15, weight: UIFontWeightRegular)
        self.previewTable.rowHeight = 30
        self.previewTable.tableFooterView = UIView(frame: CGRect.zero)
        self.showPreviewTable = true
        
        if self.parent?.view.frame.size.height == 480 {
            self.previewTable.isScrollEnabled = false
            self.showPreviewTable = false
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        assert(self.navigationController?.viewControllers.count == 4, "nav oops")
        
        let clearButton = UIBarButtonItem(title: "Clear".localized(), style: .plain, target: self, action: #selector(self.clearFiltersTapped(_:)))
        self.navigationItem.rightBarButtonItem = clearButton
    }
    
    func clearFiltersTapped(_ sender: Any) {
        self.clearFilters()
    }
    
    func clearFilters() {
        self.cardList.clearFilters()
        
        self.factionControl.selectAllSegments(false)
        self.miniFactionControl.selectAllSegments(false)
        self.typeControl.selectAllSegments(false)
        
        self.influenceSlider.value = 0.0
        self.strengthSlider.value = 0
        self.muApSlider.value = 0
        self.costSlider.value = 0
        
        self.influenceChanged(nil)
        self.strengthChanged(nil)
        self.muApChanged(nil)
        self.costChanged(nil)
        
        self.selectedTypes = Set(self.typeNames)
        self.subtypes = nil
        
        if self.showPreviewTable {
            let top = IndexPath(row: 0, section: 0)
            self.previewTable.scrollToRow(at: top, at: .middle, animated: false)
        }
    }
    
    @IBAction func strengthChanged(_ slider: UISlider?) {
        var value = Int(round(slider?.value ?? 0))
        slider?.value = Float(value)
        value -= 1
        self.strengthLabel.text = String(format: "Strength: %@".localized(), value == -1 ? "All" : String(value))
        self.cardList.filterByStrength(value)
        self.updatePreview()
    }
    
    @IBAction func costChanged(_ slider: UISlider?) {
        var value = Int(round(slider?.value ?? 0))
        slider?.value = Float(value)
        value -= 1
        self.costLabel.text = String(format: "Cost: %@".localized(), value == -1 ? "All" : String(value))
        self.cardList.filterByCost(value)
        self.updatePreview()
    }
    
    @IBAction func influenceChanged(_ slider: UISlider?) {
        var value = Int(round(slider?.value ?? 0))
        slider?.value = Float(value)
        value -= 1
        self.influenceLabel.text = String(format: "Influence: %@".localized(), value == -1 ? "All" : String(value))
        
        if let identity = self.identity {
            self.cardList.filterByInfluence(value, forFaction: identity.faction)
        } else {
            self.cardList.filterByInfluence(value)
        }
        self.updatePreview()
    }
    
    @IBAction func muApChanged(_ slider: UISlider?) {
        var value = Int(round(slider?.value ?? 0))
        slider?.value = Float(value)
        value -= 1
        let fmt = self.role == .runner ? "MU: %@".localized() : "AP: %@".localized()
        self.muApLabel.text = String(format: fmt, value == -1 ? "All" : String(value))
        
        if self.role == .runner {
            self.cardList.filterByMU(value)
        } else {
            self.cardList.filterByAgendaPoints(value)
        }
        self.updatePreview()
    }
    
    @IBAction func selectSubtype(_ btn: UIButton) {
        let picker = BrowserValuePicker(title: "Subtype".localized())
        
        var subtypes = CardManager.subtypesFor(role: self.role, andTypes: self.selectedTypes, includeIdentities: false)
        subtypes.insert(Constant.kANY, at: 0)
        
        picker.data = TableData(values: subtypes)
        
        picker.preselected = self.subtypes
        picker.setResult = { result in
            self.subtypes = result
            self.cardList.filterBySubtype(result)
            self.updatePreview()
        }
        
        self.navigationController?.pushViewController(picker, animated: true)
    }
    
    // MARK: - multi select delegate
    func multiSelect(_ control: MultiSelectSegmentedControl!, didChangeValue value: Bool, at index: UInt) {
        var set = Set<String>()
        
        if control.tag == Tags.type.rawValue {
            for idx in control.selectedSegmentIndexes {
                let type = self.typeNames[idx]
                set.insert(type)
            }
            self.selectedTypes = set
            self.subtypes = nil
            self.cardList.filterBySubtype(FilterValue.strings(Set<String>()))
            self.cardList.filterByType(FilterValue.strings(set))
        } else {
            for idx in self.factionControl.selectedSegmentIndexes {
                let faction = self.factionNames[idx]
                set.insert(faction)
            }
            if self.role == .runner {
                for idx in self.miniFactionControl.selectedSegmentIndexes {
                    let faction = self.factionNames[idx + 4]
                    set.insert(faction)
                }
            }
            
            self.cardList.filterByFaction(FilterValue.strings(set))
        }
        
        self.updatePreview()
    }
    
    func updatePreview() {
        self.cards = self.cardList.allCards()
        
        let count = self.cards.count
        let fmt = count == 1 ? "%lu matching card".localized() : "%lu matching cards".localized()
        self.previewHeader.text = "  " + String(format: fmt, count)
        
        self.previewTable.reloadData()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.showPreviewTable ? self.cards.count : 0
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let identifier = "previewCell"
        let cell = tableView.dequeueReusableCell(withIdentifier: identifier) ?? {
            let c = UITableViewCell(style: .default, reuseIdentifier: identifier)
            c.selectionStyle = .none
            c.textLabel?.font = UIFont.systemFont(ofSize: 13)
            c.textLabel?.textColor = .darkGray
            return c
        }()
        
        let card = self.cards[indexPath.row]
        cell.textLabel?.text = card.name
        
        return cell
    }
    
}
